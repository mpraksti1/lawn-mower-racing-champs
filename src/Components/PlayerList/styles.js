import styled from 'styled-components';

// TODO: refactor these styles

export const StyledTable = styled.table`
  width: 100%;
  margin: 0;
  padding: 0;
  border-collapse: collapse;
  border-spacing: 0;

  thead {
    display: none;
    color: ${({ theme }) => theme.colors.darkGray};
    background: ${({ theme }) => theme.colors.green};
    border-bottom: 2px solid ${({ theme }) => theme.colors.darkGray};
    padding: ${({ theme }) => theme.baseSizeUnit}px 0;
    
    @media screen and (min-width: 768px) {
      display: table-header-group;
    }
  }

  tr {
    border: 1px solid ${({ theme }) => theme.colors.white};
    border-bottom: 2px solid ${({ theme }) => theme.colors.white};
    background: ${({ theme }) => theme.colors.green};
    margin-bottom: 1${({ theme }) => theme.baseSizeUnit}px;
    display: block;
    position: relative;

    @media screen and (min-width: 768px) {
      display: table-row;
      border-bottom-width: 1px;
      margin-bottom: 0;
      padding: ${({ theme }) => theme.baseSizeUnit / 2}px;

      &:nth-child(odd) {
        background: ${({ theme }) => theme.colors.gray};
      }
    }
  }

  th,
  td {
    padding: ${({ theme }) => theme.baseSizeUnit}px;
    text-align: left;
  }

  th {
    text-transform: uppercase;
    font-size: ${({ theme }) => theme.fontSizes.xsm};
  }

  td {
    display: block;
    text-align: right;
    font-size: 13px;
    border-bottom: 1px dotted ${({ theme }) => theme.colors.white};
    
    &:nth-child(odd) {
      background: ${({ theme }) => theme.colors.darkGreen};
    }

    &:last-child {
      border-bottom: none;
    }

    &:before {
      content: attr(data-label);
      float: left;
      text-transform: uppercase;
      font-weight: bold;

      @media screen and (min-width: 768px) {
        content: "";
        display: none;
      }
    }

    @media screen and (min-width: 768px) {
      display: table-cell;
      text-align: left;
      font-size: ${({ theme }) => theme.fontSizes.sm};;
      border-bottom: none;
      background: none!important;
    }
  }

  .profile-pic {
    img {
      height: 20px;
      border-radius: 50%;
    }

    @media screen and (max-width: 767px) {
      text-align: center;

      img {
        height: 60px;
      }
    }
  }

  .delete-player {
    cursor: pointer;
    
    svg {
      height: 25px;
      width: 25px;
    }

    @media screen and (max-width: 767px) {
      position: absolute;
      top: ${({ theme }) => theme.baseSizeUnit}px;
      right: ${({ theme }) => theme.baseSizeUnit}px;
    }
  }
`;

StyledTable.displayName = 'PlayerListTableStyled';

export const StyledDiv = styled.div`
  background: ${({ theme }) => theme.colors.gray};
  border: 1px solid ${({ theme }) => theme.colors.green};
  padding: ${({ theme }) => theme.baseSizeUnit * 2}px;
  text-align: center;
  border-radius: 2px;
  
  button {
    margin-right: ${({ theme }) => theme.baseSizeUnit}px;

    &:last-child {
      margin-right: 0;
    }
  }
`;

StyledDiv.displayName = 'PlayerListPopUpStyled';
