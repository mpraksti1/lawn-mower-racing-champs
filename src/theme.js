const theme = {
  breakpoints: {
    mobile: 767,
    desktop: 768,
  },
  fonts: {
    baseFont: 'Helvetica',
    sansFont: 'Montserrat',
  },
  baseSizeUnit: 8,
  baseFontUnit: 16,
  colors: {
    yellow: '#f0ff0f',
    red: '#9d0404',
    gray: '#cfcfcf',
    darkGray: '#333',
    green: '#049d63',
    darkGreen: '#097149',
    white: '#eee',
    black: '#111',
    transparentBlack: 'rgba(0, 0, 0, 0.1)',
  },
  fontSizes: {
    xsm: '0.75rem',
    sm: '0.875rem',
    md: '1rem',
    lg: '1.125rem',
    xlg: '1.5rem',
    xxlg: '2.25rem',
  },
};

export default theme;
