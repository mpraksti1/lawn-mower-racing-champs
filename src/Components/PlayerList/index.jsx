import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withAlert } from 'react-alert';
import DeleteIcon from 'react-icons/lib/md/clear';
import Text from '../../Elements/Text';
import Centered from '../../Elements/Centered';
import Button from '../../Elements/Button';
import { StyledTable, StyledDiv } from './styles';

class PlayerList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      deleteModalShowing: false,
      selectedPlayer: '',
    };

    this.token = window.localStorage.token;
    this.retrievePlayers = this.retrievePlayers.bind(this);
    this.renderHeaderRow = this.renderHeaderRow.bind(this);
    this.renderRows = this.renderRows.bind(this);
    this.deleteThisPlayer = this.deleteThisPlayer.bind(this);
    this.setSelectedPlayer = this.setSelectedPlayer.bind(this);
    this.toggleDeleteModal = this.toggleDeleteModal.bind(this);
  }

  componentDidMount() {
    this.retrievePlayers();
  }

  setSelectedPlayer(evt) {
    this.toggleDeleteModal();
    this.setState({
      selectedPlayer: evt.target.getAttribute('data-player'),
    });
  }

  async retrievePlayers() {
    const { retrievePlayers, alert } = this.props;
    try {
      await retrievePlayers(this.token);
    } catch (error) {
      alert.error(error.message);
    }
  }

  async deleteThisPlayer() {
    const { deletePlayer, alert } = this.props;
    const { selectedPlayer } = this.state;

    try {
      const deleteValue = await deletePlayer(selectedPlayer, this.token);
      if (deleteValue.success) {
        this.toggleDeleteModal();
        this.retrievePlayers();
        alert.success('Racer deleted!');
      }
    } catch (error) {
      alert.error(error.message);
    }
  }

  toggleDeleteModal() {
    const { deleteModalShowing } = this.state;

    this.setState({
      deleteModalShowing: !deleteModalShowing,
    });
  }

  renderHeaderRow() {
    const { headerNames } = this.props;
    const headerRowColumns = headerNames.map((colName, i) => {
      const replacedUnderscore = colName.replace(/_/g, ' ');
      // eslint-disable-next-line
      return <th key={i}>{ replacedUnderscore }</th>;
    });

    return (
      <tr>
        <th>Profile Img</th>
        {headerRowColumns}
        <th>&nbsp;</th>
      </tr>
    );
  }

  renderRows() {
    const { headerNames, players } = this.props;

    const generateProfilePic = (i) => {
      const MW = i % 2 === 0 ? 'men' : 'women';
      return `https://randomuser.me/api/portraits/${MW}/${i + 1}.jpg`;
    };

    if (players.length > 0) {
      return players.map((player, playersIndex) => {
        if (player) {
          const columns = headerNames.map((columnName, columnIndex) => {
            const playerTrait = player[columnName];
            const label = columnName.replace(/_/g, ' ');
            // eslint-disable-next-line
            return <td key={columnIndex} data-label={label}>{playerTrait}</td>;
          });

          return (
            <tr key={player.id}>
              <td className="profile-pic">
                <img
                  src={generateProfilePic(playersIndex)}
                  alt="icon"
                />
              </td>
              {columns}
              <td className="delete-player">
                { /* Using a data attr here because cypress doesn't like being passed any
                  of the other methods of binding 'this' to the current iteration */ }
                { /* eslint-disable */ }
                <Button className="delete" data-player={player.id} islink onClick={this.setSelectedPlayer}>
                { /* eslint-enable */ }
                  <DeleteIcon />
                </Button>
              </td>
            </tr>
          );
        }
        return null;
      });
    }

    return (
      <tr>
        <td colSpan="6">
          <Text
            xlg
            thin
            sans
            spaceAround
            alignCenter
            gray
            block
          >
            {'Uh oh... you haven\'t added any racers yet.'}
          </Text>
        </td>
      </tr>
    );
  }

  render() {
    const { deleteModalShowing } = this.state;
    return (
      <div>
        <StyledTable {...this.props} className="responsive-table">
          <thead>
            {this.renderHeaderRow()}
          </thead>
          <tbody>
            {this.renderRows()}
          </tbody>
        </StyledTable>
        { deleteModalShowing && (
          <Centered fixed>
            <StyledDiv>
              <Text lg sans spaceAround block gray alignCenter>
                Are you sure you want to delete this racer?
              </Text>
              <div>
                <Button className="delete-confirm" green inline onClick={this.deleteThisPlayer}>
                  {'Yep, they\'re outta here!'}
                </Button>
                <Button outline inline onClick={this.toggleDeleteModal}>
                  Nah, I changed my mind.
                </Button>
              </div>
            </StyledDiv>
          </Centered>
        )}
      </div>
    );
  }
}

PlayerList.propTypes = {
  headerNames: PropTypes.arrayOf(PropTypes.string).isRequired,
  players: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default withAlert(PlayerList);
