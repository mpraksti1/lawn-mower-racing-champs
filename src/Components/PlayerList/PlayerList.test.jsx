import PlayerList from '.';
import theme from '../../theme';

const headerNames = ['first_name', 'last_name', 'rating', 'handedness'];
const players = [{
  first_name: 'awefff',
  last_name: 'awefawef',
  rating: 1,
  handedness: 'right',
  id: '5b5cf8b4310d217ff39015fc',
},
{
  first_name: 'awef',
  last_name: 'awfff',
  rating: 1,
  handedness: 'right',
  id: '5b5ea589310d217ff3901607',
}];

describe('The playerlist!', () => {
  it('renders without crashing', () => {
    shallow(<PlayerList />);
  });

  it('matches snapshot', () => {
    const wrapper = shallowWithTheme(
      <PlayerList headerNames={headerNames} players={players} />,
      theme,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
