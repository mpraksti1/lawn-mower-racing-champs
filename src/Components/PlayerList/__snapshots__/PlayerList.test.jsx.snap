// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The playerlist! matches snapshot 1`] = `<[object Object] />`;
